### DISCUSSION TITLE
Combining Mutation Tracking with Dataclasses or Pydantic Models

### DISCUSSION THREAD

**Discussion by connebs** [sqlalchemy/sqlalchemy#11050] [Usage Questions]
Hey, I've been struggling to try to combine Dataclasses or Pydantic with SQLA's mutation tracking capabilities, in the hopes of having a JSON column in my database that has some level of "type-safety", via said classes.

However I am really struggling to get everything working together.

Does anyone have any thoughts as to how this could be accomplished? Below is an almost complete example of what I am trying to do, I am just missing the logic that actually tracks mutations (for SQLA) within dataclasses or pydantic models.

```python
from dataclasses import asdict, dataclass

from pydantic import BaseModel
from sqlalchemy import JSON, VARCHAR, TypeDecorator, create_engine
from sqlalchemy.ext.mutable import Mutable
from sqlalchemy.orm import DeclarativeBase, Mapped, Session, mapped_column


class JsonDataclassField(TypeDecorator):
    impl = JSON

    def __init__(self, dataclass):
        self.dataclass = dataclass
        return super().__init__()

    def load_dialect_impl(self, dialect):
        return dialect.type_descriptor(JSON())

    def process_bind_param(self, value: BaseModel, _):
        return asdict(value) if value is not None else None

    def process_result_value(self, value, _):
        return self.dataclass(**value) if value is not None else None


class JsonPydanticField(TypeDecorator):
    impl = JSON

    def __init__(self, pydantic_model):
        super().__init__()
        self.pydantic_model = pydantic_model

    def load_dialect_impl(self, dialect):
        return dialect.type_descriptor(JSON())

    def process_bind_param(self, value: BaseModel, _):
        return value.model_dump() if value is not None else None

    def process_result_value(self, value, _):
        return self.pydantic_model.model_validate(value) if value is not None else None


@dataclass
class UserPrefs1:
    allow_stuff: bool = False
    count: int = 2


class UserPrefs2(BaseModel):
    allow_stuff: bool = False
    count: int = 2


# some magic class that uses `sqlalchemy.ext.mutable.Mutable` or similar
class SomeMagicClass(Mutable):
    # do magic
    pass


class Base(DeclarativeBase):
    pass


class User(Base):
    __tablename__ = "users"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(VARCHAR(50))
    prefs1: Mapped[UserPrefs1] = mapped_column(JsonDataclassField(UserPrefs1))
    prefs2: Mapped[UserPrefs2] = mapped_column(JsonPydanticField(UserPrefs2))


engine = create_engine("sqlite:///:memory:")
sess = Session(bind=engine)

Base.metadata.create_all(engine)

up1 = UserPrefs1(allow_stuff=False, count=9)
up2 = UserPrefs2(allow_stuff=True, count=3)
user = User(name="Connor", prefs1=up1, prefs2=up2)

sess.add(user)
sess.commit()

user.prefs1.count = 42
user.prefs2.count = 42
sess.commit()

user = sess.query(User).first()
if user:
    print(user.prefs1.count) # should be 42
    print(user.prefs2.count) # should be 42
```

Any help would be appreciated, have really been banging my head against the wall with this one!

**Comment by CaselIT** [comment]
Hi,

I would suggest you look at https://docs.sqlalchemy.org/en/20/orm/extensions/mutable.html#module-sqlalchemy.ext.mutable

What you are trying to do is the same as trying to modify a dict in-place. That will not work unless you a mutable dict.
It should be similar when using classes instead of a dict.

**Reply by connebs** [reply]
Yep, that's where I started. Unfortunately was struggling to bridge the gap between the relatively simple `dict` example and the moderately more complex requirements to do this with pydantic or dataclasses. Mostly because dataclasses aren't actually class-based (inheritance-wise) and pydantic models have some underlying complexity that was making it hard to sync everything.

**Reply by connebs** [reply]
For example this is the first-thing I tried:

```python
class MutableModel(Mutable, BaseModel):
    @classmethod
    def coerce(cls, key, value):
        if not isinstance(value, MutableModel):
            if isinstance(value, BaseModel):
                return MutableModel(**value.model_dump())
            if isinstance(value, dict):
                return MutableModel(**value)

            # this call will raise ValueError
            return Mutable.coerce(key, value)
        else:
            return value

    def __setattr__(self, key, value):
        super().__setattr__(key, value)
        self.changed()

    # def __delattr__(self, key):
    #     super().__delattr__(key)
    #     self.changed()
    
    
class User(Base):
    ...
    prefs2: Mapped[UserPrefs2] = mapped_column(MutableModel.as_mutable(JsonPydanticField(UserPrefs2)))
```

But of course when you try to run the code with that, you'll get a `ValueError: "MutableModel" object has no field "count"`, because `UserPrefs2` is a subclass of `BaseModel`, not a `BaseModel` itself. `dict`s are a bit more interchangeable.

You can of course just use your `UserPrefs2` model throughout and then it works reasonably well, but I was hoping for a more generic solution that would allow encapsulation. To that end I tried things like making the above class configurable with an `__init__` dunder that you can pass a Pydantic model to, but since `coerce` is a class method that didn't go so well either.

**Reply by CaselIT** [reply]
Regarding pydantic, don't know. It's all kind of unhelpful if you try to do thing even a bit differently than what they expect, so I would not be surprised if you can't use mutable with pydantic.

But I think you should be able to create a dataclass of a mutable class by doing something like
```py
@dataclass
class MyClass(Mutable):
  ...
```

**Comment by connebs** [comment]
Ok so after a little more thought, I did realize that I can do this:

```python
def MutableModel(pydantic_model):
    class MutableModel(Mutable, pydantic_model):
        @classmethod
        def coerce(cls, key, value):
            if not isinstance(value, MutableModel):
                if isinstance(value, pydantic_model):
                    return MutableModel(**value.model_dump())
                if isinstance(value, dict):
                    return MutableModel(**value)

                # this call will raise ValueError
                return Mutable.coerce(key, value)
            else:
                return value

        def __setattr__(self, key, value):
            super().__setattr__(key, value)
            self.changed()

    return MutableModel.as_mutable(JsonPydanticField(pydantic_model))

class User(Base):
    __tablename__ = "users"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(VARCHAR(50))
    prefs: Mapped[UserPrefs] = mapped_column(MutableModel(UserPrefs))
```

However something feels very foot-gunny about this, though without better knowledge of SQLA internals I can't say what those might be.

**Comment by seanabraham** [comment]
@acnebs how is that most recent implementation working out for you? Run into any issues with it?

**Comment by joakimnordling** [comment]
I ran into the same problem and ended up publishing a small library for it: [sqlalchemy-pydantic-json](https://github.com/joakimnordling/sqlalchemy-pydantic-json) (I'm the author).

It starts from the same idea as the `Mutable` + `BaseModel` combination above, but also tracks changes inside the model: nested models, lists, dicts and sets, at any depth. So `user.prefs.tags.append("x")` or `user.prefs.address.city = "Espoo"` marks the row as changed, without `flag_modified()`:

```python
from sqlalchemy_pydantic_json import EmbeddedPydanticModel

class UserPrefs(EmbeddedPydanticModel):
    allow_stuff: bool = False
    count: int = 2
    tags: list[str] = []

class User(Base):
    ...
    prefs: Mapped[UserPrefs] = mapped_column(UserPrefs.column(), default=UserPrefs)
```

It's tested on SQLite, PostgreSQL and MariaDB with SQLAlchemy 2.0 and 2.1, type-checks with mypy, pyright and ty, and has Alembic support. It's however only for Pydantic. Would be really nice to get some feedback on it; feel free to submit issues if you find something missing or have some improvement ideas or find any bugs.

**Reply by zzzeek** [reply]
cool!   I've only looked at pydantic very slightly, enough to observe getting SQLAlchemy to work with it would be very difficult.   But I welcome tools that try to smooth the gaps.
